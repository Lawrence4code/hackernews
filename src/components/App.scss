.header {
  display: grid;
  grid-template-columns: 35% 65%;
  margin-bottom: 2rem;
  text-align: left;

  &__logo {
    color: #102974;
    font-size: 1.5rem;
    text-transform: uppercase;
  }

  &__search {
    text-align: left;

    & label {
      opacity: 0;
    }

    & input {
      color: #555;
      width: 25%;
    }

    & button {
      background-color: #fff;
    }

    & button:hover {
      color: rgb(31, 66, 168);
      border: 1px solid rgba(65, 105, 225, 0.753);
    }
  }
}

.page {
  margin: 2% 10% 0 10%;
}

.interactions {
  text-align: center;
}

.table {
  margin: 20px 0;
}

.table-header {
  display: flex;
  font-size: 16px;
  justify-content: space-between;
  line-height: 24px;
  margin-bottom: 1rem;
  padding: 0 10px;
}

.table-empty {
  font-size: 16px;
  margin: 200px;
  text-align: center;
}

.table-row {
  background: #ffffff;
  border: 1px solid #e3e3e3;
  display: flex;
  line-height: 24px;
  margin: 10px 0;
  padding: 10px;
  white-space: nowrap;
}

.table-header {
  width: 100%;
  display: grid;
  grid-template-columns: (50% 20% 10% 10% 10%);
}

.table-header > span {
  color: #5f5f5f;
  overflow: hidden;
  padding: 0 5px;
  text-overflow: ellipsis;

  &:nth-child(3),
  &:nth-child(4),
  &:nth-child(5) {
    text-align: center;
  }
}

.table-row {
  width: 100%;
  display: grid;
  grid-template-columns: (50% 20% 10% 10% 10%);
}

.table-row > span {
  overflow: hidden;
  padding: 0 5px;
  text-overflow: ellipsis;

  &:nth-child(3),
  &:nth-child(4),
  &:nth-child(5) {
    text-align: center;
  }
}

.button-inline {
  border-width: 0;
  background: transparent;
  color: inherit;
  cursor: pointer;
  font-size: inherit;
  -webkit-font-smoothing: inherit;
  padding: 0;
  text-align: inherit;
}

.button-active {
  color: #4169e1;
}

footer {
  margin-top: 0.5rem;
  padding-bottom: 0.5rem;
  text-align: right;

  bottom: 10px;

  a {
    color: rgb(31, 62, 155);
    text-decoration: none;
  }
}

.post-dismiss {
  transition: color 0.4s ease;
  &:hover {
    border: none;
    color: orangered;
  }
}

.post-title {
  text-decoration: none;
  font-size: 0.925rem;
  transition: color 0.4s ease;
  &:hover {
    color: #4169e1;
  }
}

.more-btn {
  background-color: #fff;
  &:hover {
    color: rgb(31, 66, 168);
    border: 1px solid rgba(65, 105, 225, 0.753);
  }
}
