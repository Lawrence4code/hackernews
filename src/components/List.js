import React from 'react';
import Button from './Button';
import propTypes from 'prop-types';
import { sortBy } from 'lodash';

const SORTS = {
  NONE: list => list,
  TITLE: list => sortBy(list, 'title'),
  AUTHOR: list => sortBy(list, 'author'),
  COMMENTS: list => sortBy(list, 'num_comments').reverse(),
  POINTS: list => sortBy(list, 'points').reverse()
};

const List = ({ list, sortKey, onSort, onDismiss }) => (
  <div className="list">
    {SORTS[sortKey](list).map(item => (
      <div key={item.objectID} className="list-row">
        <span style={{ width: '40%' }}>
          <a href={item.url}> {item.title || 'No Title'}</a>
        </span>
        <span style={{ width: '30%' }}>{item.author}</span>
        <span style={{ width: '10%' }}>{item.num_comments}</span>
        <span style={{ width: '10%' }}>{item.points}</span>
        <span style={{ width: '10%' }}>
          <Button
            onClick={() => {
              onDismiss(item.objectID);
            }}
            className="button-inline"
          >
            Dismiss
          </Button>
        </span>
      </div>
    ))}
  </div>
);

Button.propTypes = {
  onClick: propTypes.func.isRequired,
  className: propTypes.string,
  children: propTypes.node.isRequired
};

export default List;
