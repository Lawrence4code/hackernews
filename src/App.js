import React, { Component } from 'react';
import Search from './components/Search';
import List from './components/List';
import './App.css';
import Button from './components/Button';

// const list = [
//   {
//     title: 'React',
//     url: 'https://reactjs.org',
//     author: 'Jordan Walke',
//     num_comments: 3,
//     points: 4,
//     objectId: 0
//   },
//   {
//     title: 'Redux',
//     url: 'https://redux.org/',
//     author: ['Dan Abramov', 'Andrew Clark'],
//     num_comments: 2,
//     points: 5,
//     objectId: 1
//   }
// ];

const DEFAULT_QUERY = 'redux';
const DEFAULT_HPP = '12';
const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';
const PARAM_HPP = 'hitsPerPage=';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = { list: null, searchTerm: DEFAULT_QUERY };
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
  }

  setSearchTopStories(list) {
    const { hits, page } = list;
    const oldHits = page !== 0 ? this.state.list.hits : [];
    const updatedHits = [...oldHits, ...hits];
    this.setState(() => {
      return { list: { hits: updatedHits, page } };
    });
  }

  fetchSearchTopStories(searchTerm, page = 0) {
    console.log('fetchSearchTopStories triggered');

    fetch(
      `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`
    )
      .then(response => response.json())
      .then(list => this.setSearchTopStories(list))
      .catch(error => error);
  }

  componentDidMount() {
    console.log(this.state);
    const { searchTerm } = this.state;
    this.fetchSearchTopStories(searchTerm);
  }

  onSearchSubmit(event) {
    console.log('onSearchSubmit triggered.');
    event.preventDefault();
    const { searchTerm } = this.state;
    this.fetchSearchTopStories(searchTerm);
  }

  onDismiss(id) {
    const isNotId = item => item.objectID !== id;
    const newList = this.state.list.hits.filter(isNotId);
    this.setState(() => {
      return {
        list: {
          ...this.state.list,
          hits: newList
        }
      };
    });
  }

  onSearchChange(event) {
    this.setState({ searchTerm: event.target.value });
  }

  render() {
    // object destructuring
    const { list, searchTerm } = this.state;
    console.log(this.state);
    const page = (list && list.page) || 0;
    // if no data in list, return null
    if (!list) {
      return null;
    }

    // filtering the list based on search input and then mapping over it to render filtered list (refactored and removed , to ensure server side fitlering)
    // const filteredList = list.hits.filter(item => {
    //   return item.title.toLowerCase().includes(searchTerm.toLowerCase());
    // });

    return (
      <div className="page">
        <div className="interactions">
          {/* Search only handles the search term event */}
          <Search
            value={searchTerm}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
          >
            {/* Passing Search text as child to this component which can be access from this.props in Search component */}
            Search
          </Search>
          {list && <List list={list.hits} onDismiss={this.onDismiss} />}
          <Button
            onClick={() => this.fetchSearchTopStories(searchTerm, page + 1)}
          >
            More
          </Button>
        </div>
      </div>
    );
  }
}

export default App;
