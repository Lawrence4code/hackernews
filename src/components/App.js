import React, { Component } from 'react';
import axios from 'axios';
import Search from './Search';
import List from './List';
import './App.css';
import Button from './Button';
import propTypes from 'prop-types';

import pageLoading from '../assets/images/post-loader.gif';

import {
  DEFAULT_QUERY,
  DEFAULT_HPP,
  PATH_BASE,
  PATH_SEARCH,
  PARAM_SEARCH,
  PARAM_PAGE,
  PARAM_HPP
} from '../constants';

const PageLoading = () => {
  return (
    <div>
      <img src={pageLoading} alt="page loading" />
    </div>
  );
};

const PostLoading = () => {
  return (
    <div>
      <img src={pageLoading} alt="post loading" />
    </div>
  );
};

const withLoading = Component => ({ isLoading, ...rest }) =>
  isLoading ? <PostLoading /> : <Component {...rest} />;

const ButtonWithLoading = withLoading(Button);

class App extends Component {
  _isMounted = false;
  constructor(props) {
    super(props);
    this.state = {
      lists: null,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      error: null,
      isLoading: false,
      sortKey: 'NONE'
    };
    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
    this.onSort = this.onSort.bind(this);
  }

  setSearchTopStories(list) {
    const { hits, page } = list;
    const { searchKey, lists } = this.state;
    const oldHits = lists && lists[searchKey] ? lists[searchKey].hits : [];
    const updatedHits = [...oldHits, ...hits];
    this.setState(() => {
      return {
        lists: { ...lists, [searchKey]: { hits: updatedHits, page } },
        isLoading: false
      };
    });
  }

  fetchSearchTopStories(searchTerm, page = 0) {
    this.setState(() => {
      return { isLoading: true };
    });
    axios(
      `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`
    )
      .then(list => this._isMounted && this.setSearchTopStories(list.data))
      .catch(
        error =>
          this._isMounted &&
          this.setState(() => {
            return { error };
          })
      );
  }

  needsToSearchTopStories(searchTerm) {
    return !this.state.lists[searchTerm];
  }

  componentDidMount() {
    this._isMounted = true;
    const { searchTerm } = this.state;
    this.setState(() => {
      return {
        searchKey: searchTerm
      };
    });
    this.fetchSearchTopStories(searchTerm);
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  onSearchSubmit(event) {
    event.preventDefault();
    console.log('onSearchSubmit triggered.');
    const { searchTerm } = this.state;
    this.setState(() => {
      return {
        searchKey: searchTerm
      };
    });

    if (this.needsToSearchTopStories(searchTerm)) {
      console.log(searchTerm);
      this.fetchSearchTopStories(searchTerm);
    }
  }

  onDismiss(id) {
    const { lists, searchKey } = this.state;
    const { hits, page } = lists[searchKey];
    const isNotId = item => item.objectID !== id;
    const newList = hits.filter(isNotId);
    this.setState(() => {
      return {
        lists: {
          ...lists,
          [searchKey]: { hits: newList, page }
        }
      };
    });
  }

  onSearchChange(event) {
    this.setState({ searchTerm: event.target.value });
  }

  onSort(sortKey) {
    this.setState(sortKey => {
      return {
        sortKey
      };
    });
  }

  render() {
    // object destructuring
    const {
      lists,
      searchKey,
      searchTerm,
      error,
      isLoading,
      sortKey
    } = this.state;
    const page = (lists && lists[searchKey] && lists[searchKey].page) || 0;
    const list = (lists && lists[searchKey] && lists[searchKey].hits) || [];
    // if no data in list, return null
    // if (!list) {
    //   return null;
    // }

    // filtering the list based on search input and then mapping over it to render filtered list (refactored and removed , to ensure server side fitlering)
    // const filteredList = list.hits.filter(item => {
    //   return item.title.toLowerCase().includes(searchTerm.toLowerCase());
    // });

    return (
      <div className="page">
        <div className="interactions">
          {/* Search only handles the search term event */}
          <Search
            value={searchTerm}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
          >
            {/* Passing Search text as child to this component which can be access from this.props in Search component */}
            Search
          </Search>

          {error ? (
            <p>
              Something went wrong. Possible issue with URL or type of search,
              please try again or report the issue.
            </p>
          ) : (
            <div>
              <List
                sortKey={sortKey}
                onSort={this.onSort}
                list={list}
                onDismiss={this.onDismiss}
              />
              <div className="interactions">
                {isLoading ? (
                  <PageLoading />
                ) : (
                  <ButtonWithLoading
                    isLoading={isLoading}
                    onClick={() =>
                      this.fetchSearchTopStories(searchKey, page + 1)
                    }
                  >
                    More
                  </ButtonWithLoading>
                )}
              </div>
            </div>
          )}
        </div>
      </div>
    );
  }
}

Button.propTypes = {
  onClick: propTypes.func.isRequired,
  children: propTypes.node.isRequired
};

export default App;
