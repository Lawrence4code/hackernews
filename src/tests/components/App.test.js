import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import App from '../../components/App';
import List from '../../components/List';

describe('App Component', () => {
  test('render App without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('should render App Component correctly', () => {
    const component = renderer.create(<App />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
