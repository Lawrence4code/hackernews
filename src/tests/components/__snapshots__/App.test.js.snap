// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should render App Component correctly 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value="redux"
      />
      <button
        type="submit"
      >
         
        Search
         
      </button>
    </form>
    <div>
      <div
        className="list"
      />
      <div
        className="interactions"
      >
        <div>
          <img
            alt="page loading"
            src="post-loader.gif"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
